penguin/tools/schema_contract.py

```
@@ -0,0 +1,155 @@
+
+def render_tool_usage_guidance(tool_schema: Dict[str, Any]) -> str:
+    """Render concise model-facing usage guidance from a tool schema."""
+
+    name = str(tool_schema.get("name") or "tool").strip() or "tool"
+    input_schema = tool_schema.get("input_schema")
+    if not isinstance(input_schema, dict):
+        input_schema = {"type": "object", "properties": {}}
+
+    properties = input_schema.get("properties")
+    property_names = (
+        sorted(str(key) for key in properties.keys())
```

Ruff (0.15.12)

[warning] Use `key in dict` instead of `key in dict.keys()`

Remove `.keys()`

(SIM118)